provider/cluster/hostname.go

```diff
@@ -126,25 +266,47 @@ loop:
 			break loop
 		case rr := <-hs.requests:
 			hs.doRequest(rr)
-		case hostnames := <-hs.releases:
-			hs.doRelease(hostnames)
+		case crr := <-hs.canRequest:
+			canReserveHostnamesImpl(hs.inUse, crr.hostnames, crr.ownerAddr, crr.result)
+		case dID := <-hs.releases:
+			releaseHostnamesImpl(hs.inUse, dID)
+		case request := <-hs.prepareRequest:
+			prepareHostnamesImpl(hs.inUse, request.hostnames, request.deploymentID, request.chErr)
+
 		}
 	}
 
 }
 
-var errHostnameNotAllowed = errors.New("hostname not allowed")
+var ErrHostnameNotAllowed = errors.New("hostname not allowed")
+
+func (hs *hostnameService) PrepareHostnamesForTransfer(hostnames []string, dID dtypes.DeploymentID) <- chan error {
+	chErr := make(chan error, 1)
+
+	v:= prepareTransferRequest{
+		hostnames: hostnames,
+		deploymentID: dID,
+		chErr:     chErr,
+	}
+	select {
+		case hs.prepareRequest <- v:
+	case <-hs.lc.ShuttingDown():
+		chErr <- ErrNotRunning
+	}
+
+	return chErr
```

Comment:
It is generally cleaner to contain asynchronous things within the implementation here, like is done with everything else.  Why are we returning a channel here?

Reply:
I will see about refactoring this